### PR TITLE
Add support for GraphQL Argument prepare methods when defined as a method on the constant

### PR DESCRIPTION
**Pull request by DougEdey**
### Motivation
Second part of https://github.com/Shopify/tapioca/issues/1717

This adds support for prepare method when defined as a method on the constant.

[It uses the same check as GraphQL-Ruby](https://github.com/rmosolgo/graphql-ruby/blob/109c26c5fa9118a9c69ed4c4823048c1ce1aea6d/lib/graphql/schema/argument.rb#L217-L226)

And I use the same fallback behaviour as I did in #1720, where it falls back to the base type if the signature is missing or void (technically this is a bad signature I guess? But I'm not sure what we should really be doing here)

### Implementation

I added `constant` as a param in `Tapioca::Dsl::Helpers::GraphqlTypeHelper#type_for` since that holds the actual method on it, and as such I had to update `GraphqlInputObject` to use it too.

### Tests

Yes!

### CODE REVIEW

**Comment by eapache-opslevel** [comment]
> technically this is a bad signature I guess? But I'm not sure what we should really be doing here
Eh, many `prepare` methods do transforms that don't change the type. I think this is the correct / best-effort guess here.

**Comment by DougEdey** [comment]
@Morriar @st0012 👋 this is "As good as it's gonna get" for now, I can't find a way to support the `owner defined prepare method` because the following syntax is valid:

```ruby
class MyMutation < GraphQL::Schema::Mutation
  argument :foo, ID, required: true, prepare: :some_method
 
end

class MutationRoot < GraphQL::Schema::Object
  field :my_mutation, mutation: MyMutation

  class << self
    extend T::Sig

    sig { params(in: T.untyped).returns(Integer) }
    def some_method(in)
      in.to_i
    end
  end
end

class OtherMutationRoot < GraphQL::Schema::Object
  field :my_mutation, mutation: MyMutation

  class << self
    extend T::Sig

    sig { params(in: T.untyped).returns(String) }
    def some_method(in)
      in.to_s
    end
  end
end
```

And we can't statically detect what the caller is, it's a weird syntax to use anyway, and the changes I've got in the MR for GraphQL Ruby will definitely improve the situation to make this much clearer

**Comment by DougEdey** [comment]
@Morriar It turns out that syntax isn't really supported and is legacy, it doesn't make any sense for the actual Gem to work that way and wouldn't work with static typing. So this MR is as good as it'll get because the example I gave is only known at Runtime and can't be statically typed

File: spec/tapioca/dsl/compilers/graphql_mutation_spec.rb [hunk 1]
```diff
@@ -172,7 +172,51 @@ def resolve(boolean:, float:, id:, int:, big_int:, date:, datetime:, json:, stri
...
+                    end
+
+                    def prepare_dates_untyped(other, _context)
+                      other
+                    end
+                  end
+
+                  argument :min, GraphQL::Types::ISO8601Date, "Minimum value of the range", prepare: :prepare_dates
+                  argument :max, GraphQL::Types::ISO8601Date, "Maximum value of the range" #, prepare: prepare_dates_void
+                  argument :other, GraphQL::Types::ISO8601Date, "Some value of the range " #, prepare: prepare_dates_untyped
```
[the remark this answers is not in the document]
Comment (DougEdey) [reply spec/tapioca/dsl/compilers/graphql_mutation_spec.rb]: Oh man I am so sorry! I forgot to push a commit!

**Pull request merged**